cityofaustin/atd-airflow
# DAG to invoke CRIS import script

## Associated PRs

This PR has a sibling PR in the VZ repository which contains the image that is invoked in this DAG. You can find it here: https://github.com/cityofaustin/atd-vz-data/pull/1235.



## Testing

We have a few options here. This can be a review the code and rely on my testing, or I would be more than happy to give a demonstration of this working on my local VZ & airflow stacks. I can do that individually or maybe even as a dev-sync share if that would be helpful. 

Of course, individual testing is always welcome! Here is the general outline of how to do it. There are shared resources in here like the SFTP endpoint, so you may want to mention that you're doing this testing in the dev channel or something so two people don't try to do this at the same time.

* Modify your local VZ stack to expose the `graphql-engine` on port 8084, because 8080 will be used by the airflow web ui.
* Spin up the local VZ DB and populate it with data from the read replica.
* Spin up the local `graphql-engine` instance, now listening on 8084 of your machine.
* Bring up the local airflow stack on this branch. 
  * We have some other DAGs in flux in terms of what they are expecting in 1PW and what is available, so I had to remove them from my local checkout so I didn't get import errors in the airflow UI.
* Get on the SFTP endpoint and use `sudo` to become root so you can jump into `txdot`'s home directory. See that it is empty. If it's not, you're probably at work really early (or up after midnight working late) and looking at a file that will be imported that day at 7AM. This file is important, and it's probably easiest if you leave it in place until the ETL consumes it at 7AM.
* Sign into Chia's CRIS account, and request an export be uploaded. From memory, the parameters you need are:
  * CSV export
  * Process date (not crash date)
  * Tri-county area (Wilco, Travis & Hays)
  * Extract password out of 1PW
  * A process date interval of your choosing, but it must be entirely in the past and not include today. The longer the interval you pick, the longer your testing is going to take BTW. A day or two is good.
* Watch that `/home/txdot` folder with something like `while sleep 5; do ls -l; done` and wait for the file to be delivered - normally 15-30 minutes.
* And finally, you can trigger the DAG. 
  * It'll pull the import image of docker hub, and the image is built with both X86 and ARM chip support, so this should work natively on modern macs and linux machines. 
  * If you are on a linux machine though, you'll need to make some changes in the 1PW entry or docker-compose file that remove the reliance on the use of the magic DNS entry provided by Docker Desktop: `host.docker.internal`.
* You should see the DAG complete, and you should have updated data in your database and you also should see the file removed from the SFTP endpoint.

## ✅ ~Merge Warning~

This branch has had both the [md-12432-1pw-util](https://github.com/cityofaustin/atd-airflow/pull/107) and the [md-12363-slack-notifier](https://github.com/cityofaustin/atd-airflow/pull/102) branch merged into it.  It should not go in until both of those PRs have been merged.

- [x] Slack Notifier PR
- [x] 1Password Utility PR
